TACC/Core-Portal
# task/TUP-1747: TUP project systems

## Overview

Onboarding and datafiles support for provisioning and displaying TAPIS systems based on TAS Projects, with configurability inside the admin console

## Related

* [FP-1747](https://jira.tacc.utexas.edu/browse/FP-1747)

## Changes

- `tas_project_systems` Django application
- `tas_project_systems_step` onboarding step
- New sample configurations for `PORTAL_TAS_PROJECT_SYSTEMS_TEMPLATES`
- Datafiles attempts to list TAS Project Systems if they are configured
- Management command for forcing tas project system creation per user


## Testing

1. Make sure you are on the A-ccsc project (sql id 52) and TACC-Team project (sql id 23881). Most staff should be on these. Also, on a different portal make sure you stockyard work directory has a `bcbs` subdirectory and a `submissions` subdirectory. These will be demonstration locations for this test.
2. Use these `settings_custom` settings:

```

#################################
# DJANGO APP: TAS_PROJECT_SYSTEMS 
#################################

_PORTAL_TAS_PROJECT_SYSTEMS_TEMPLATES = {
    'apcd-organization': {
        'name': '{projectname} (APCD)',
        'description': 'Organizational storage for {projectname} (APCD)',
        'site': 'cep',
        'systemId': 'apcd-test.{projectid}.{username}',
        'host': 'cloud.corral.tacc.utexas.edu',
        'rootDir': '/work/{tasdir}/{projectdir}',
        'port': 2222,
        'icon': None,
        'hidden': False,
    },
    'apcd-submissions': {
        'name': 'Submissions (APCD)',
        'description': 'Submission storage for (APCD)',
        'site': 'cep',
        'systemId': 'apcd-test.submissions.{username}',
        'host': 'cloud.corral.tacc.utexas.edu',
        'rootDir': '/work/{tasdir}/{projectdir}',
        'port': 2222,
        'icon': None,
        'hidden': False,
    }
}

########################
# DJANGO APP: ONBOARDING
########################

_PORTAL_USER_ACCOUNT_SETUP_STEPS = [
    {
        'step': 'portal.apps.onboarding.steps.project_membership.ProjectMembershipStep',
        'settings': {
            'project_sql_id': 52,
            'rt_queue': 'Accounting'     # Defaults to 'Accounting' if left blank
        }
    },
    {
        'step': 'portal.apps.onboarding.steps.mfa.MFAStep',
        'settings': {}
    },
    {
        'step': 'portal.apps.onboarding.steps.allocation.AllocationStep',
        'settings': {}
    },
    {
        'step': 'portal.apps.onboarding.steps.key_service_creation.KeyServiceCreationStep',
        'settings': {}
    },
    {
        'step': 'portal.apps.onboarding.steps.tas_project_system_creation.TasProjectSystemCreationStep',
        'settings': {}
    }
]

```
3. `make build && make start`
4. Apply migrations and setup the new tas_project_systems elasticsearch index:

```
from portal.libs.elasticsearch.indexes setup_tas_project_systems_index
setup_tas_project_systems_index(force=True)
```

5. Make yourself superuser:

```
from django.contrib.auth import get_user_model
myuser = get_user_model().objects.get(username="myuser")
myuser.is_staff = True
myuser.is_superuser = True
myuser.save()
```

6. Add a TasProjectSystemEntry for the submissions system, by going to https://cep.dev/core/admin and adding an entry for:

```
project_sql_id: 52
projectname: Submissions
projectid: submissions
projectdir: submissions
template: apcd-submissions
```

8. The management interface will automatically create a system called "Submissions (APCD)" for Datafiles. It will take a few seconds for this to appear.  See that the new systems are listed in DataFiles and that they are accessible.
9. To test system creation, we will skip the admin management interface and use the python shell to create a TasProjectSystemEntry. (This will bypass automatic creation.)

```
from portal.apps.tas_project_systems.models import TasProjectSystemEntry
entry = TasProjectSystemEntry(project_sql_id=23881, projectid="bcbs", projectname="Blue Cross", projectdir="bcbs", template="apcd-organization")
entry.save()
```
10. **Make sure you have ngrok setup**. Use the Onboarding Admin interface to reset the TAS Project Creation step. It should see that there is a new TasProjectSystemEntry and create a system for Datafiles called "BCBS (APCD)"
11. Try doing something like changing the configuration for the `apcd-submissions` template in settings_custom (such as the description template). Shutdown the containers and restart. Now try applying the settings via management command: 

```
python manage.py force_tas_project_system_creation -u jchuah -p 52
```

13. Try changing one of the display names of the projects in the admin console. This should kick off a task to rewrite the system definition.

## UI

<img width="203" alt="image" src="https://user-images.githubusercontent.com/17181582/181598976-0f36e58d-1b53-4404-8305-d73edd5129d0.png">

<img width="535" alt="image" src="https://user-images.githubusercontent.com/17181582/181599123-38d5f7a3-4f70-4e77-b4ca-446edb2d17fa.png">

<img width="552" alt="image" src="https://user-images.githubusercontent.com/17181582/181599198-60096ff8-3f45-43f7-b487-4fa3bc4a2faf.png">


## Notes

Theory of operation:

The PORTAL_TAS_PROJECT_SYSTEMS_TEMPLATES support a dynamic `{projectname}`, `{projectdir}` and `{projectdir}` variable that can be used in system names and descriptions in those templates. They are dynamically substituted in during data files listing and tas_project_systems_creation_step based on a user's current TAS projects. If you add or modify a TAS Project in the Admin , it will find the appropriate template and substitute in those values and any user that currently has that project will have a system created or modified to fit those templates. These systems will then be displayed in their Data Files. (After the first processing or viewing of Data Files, the system variables are cached in elasticsearch for faster retrieval)
